// html-validate/build/shim.js
"use strict";
/* used when calling require('htmlvalidate'); */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.version = exports.Parser = exports.TemplateExtractor = exports.Reporter = exports.Rule = exports.UserError = exports.TextNode = exports.NodeClosed = exports.HtmlElement = exports.DynamicValue = exports.Severity = exports.ConfigLoader = exports.ConfigError = exports.Config = exports.CLI = exports.HtmlValidate = void 0;
var htmlvalidate_1 = require("./htmlvalidate");
Object.defineProperty(exports, "HtmlValidate", { enumerable: true, get: function () { return __importDefault(htmlvalidate_1).default; } });
var cli_1 = require("./cli/cli");
Object.defineProperty(exports, "CLI", { enumerable: true, get: function () { return cli_1.CLI; } });
var config_1 = require("./config");
Object.defineProperty(exports, "Config", { enumerable: true, get: function () { return config_1.Config; } });
Object.defineProperty(exports, "ConfigError", { enumerable: true, get: function () { return config_1.ConfigError; } });
Object.defineProperty(exports, "ConfigLoader", { enumerable: true, get: function () { return config_1.ConfigLoader; } });
Object.defineProperty(exports, "Severity", { enumerable: true, get: function () { return config_1.Severity; } });
var dom_1 = require("./dom");
Object.defineProperty(exports, "DynamicValue", { enumerable: true, get: function () { return dom_1.DynamicValue; } });
Object.defineProperty(exports, "HtmlElement", { enumerable: true, get: function () { return dom_1.HtmlElement; } });
Object.defineProperty(exports, "NodeClosed", { enumerable: true, get: function () { return dom_1.NodeClosed; } });
Object.defineProperty(exports, "TextNode", { enumerable: true, get: function () { return dom_1.TextNode; } });
var error_1 = require("./error");
Object.defineProperty(exports, "UserError", { enumerable: true, get: function () { return error_1.UserError; } });
var rule_1 = require("./rule");
Object.defineProperty(exports, "Rule", { enumerable: true, get: function () { return rule_1.Rule; } });
var reporter_1 = require("./reporter");
Object.defineProperty(exports, "Reporter", { enumerable: true, get: function () { return reporter_1.Reporter; } });
var transform_1 = require("./transform");
Object.defineProperty(exports, "TemplateExtractor", { enumerable: true, get: function () { return transform_1.TemplateExtractor; } });
var parser_1 = require("./parser");
Object.defineProperty(exports, "Parser", { enumerable: true, get: function () { return parser_1.Parser; } });
const pkg = require("../package.json");
exports.version = pkg.version;
